Fix NatsUri not reflecting host/port changes from OnConnectingAsync hook

tormoder opened the pull request:
Fixes https://github.com/nats-io/nats.net/issues/996.

tormoder committed:
Fix NatsUri not reflecting host/port changes from OnConnectingAsync hook

mtmk commented:
Thanks @tormoder good catch! this is fixing some of the logging but exceptions are still coming through with the original URI since the fix is done inside `ConnectSocketAsync(uri)` call and try/catch blocks wrapping it are still using the old uri. ~~Maybe it can return the uri?~~ **edit:** that won't work if it's throwing

mtmk commented on src/NATS.Client.Core/NatsConnection.cs:
this one is still reporting the original uris

mtmk commented on src/NATS.Client.Core/NatsConnection.cs:
this is still reporting the original

mtmk commented on src/NATS.Client.Core/NatsConnection.cs:
this is fine

mtmk commented on src/NATS.Client.Core/NatsConnection.cs:
maybe this could return the uri (actually that won't work with exception!)

mtmk commented on src/NATS.Client.Core/NatsConnection.cs:
```diff
@@ -452,7 +452,8 @@ private async Task ConnectSocketAsync(NatsUri uri)
             target = await OnConnectingAsync(target).ConfigureAwait(false);
             if (target.Host != uri.Host || target.Port != uri.Port)
             {
-                uri = uri with { Uri = new UriBuilder(uri.Uri) { Host = target.Host, Port = target.Port, }.Uri };
+                var modifiedUri = new UriBuilder(uri.Uri) { Host = target.Host, Port = target.Port }.Uri;
+                uri = new NatsUri(modifiedUri.ToString(), uri.IsSeed, uri.Uri.Scheme);
             }
         }
```
maybe this can be extracted into a method?

mtmk reviewed (commented):
thanks @tormoder, I saw a couple of places exceptions were still reporting the original uri

tormoder commented:
Thanks, I will take another look.

tormoder committed:
Preserve and log resolved URI after OnConnectingAsync hook

tormoder commented:
A bit fiddly, but did another attempt.

mtmk commented:
> A bit fiddly, but did another attempt.

thanks @tormoder yes it looks a bit fiddly :) it seems to work though. I will have a closer look once the other PR is merged that fixes the tests as well.

mtmk committed:
Merge branch 'main' into fork/tormoder/connect-socket-fix-uri-assignment-on-override

mtmk committed:
Merge branch 'main' into fork/tormoder/connect-socket-fix-uri-assignment-on-override

mtmk committed:
Fix NatsUri redaction

No need to over optimize redacted value it's never on hot path.
Correcness first.

mtmk commented:
hey @tormoder it looks like the problem was the NatsUri redaction caching. it's never on a hot path so just removed the caching of that string. That seems to be fixing the issue, let me know it works for you too.

tormoder commented:
> hey @tormoder it looks like the problem was the NatsUri redaction caching. it's never on a hot path so just removed the caching of that string. That seems to be fixing the issue, let me know it works for you too.

Yes, thank you.

scottf reviewed (approved):
LGTM
